Document how to run web-auth-endpoint locally

karikushi opened the pull request:
Related to Issue 215 in the Interbit repo (https://github.com/interbit/interbit/issues/215).

- added a doc for setting up and running `platform-deploy` locally
- added a doc for setting up and running `web-auth-endpoint` locally
- referenced very briefly that the GitHub OAuth loop on the accounts app works locally via these steps

I consider this a skeleton/draft, so any input on how to improve this would be appreciated.

Commit:
Add example docs for platform-deploy and web-auth-endpoint packages

Commit:
Enclose localhost links in back ticks

Commit:
Fix typos

nicholeous commented on examples/platform-deploy.md:
```diff
@@ -0,0 +1,60 @@
+# Deploying an Interbit Platform
+
+In the `packages/platform-deploy` directory, we demonstrate how to run 
+multiple Interbit applications connected to a single back-end node. This 
```
single node that runs many chains

nicholeous commented on examples/platform-deploy.md:
```diff
@@ -0,0 +1,60 @@
+multiple Interbit applications connected to a single back-end node. This 
+example generates an Interbit manifest file and covenant files for the 
+`packages/app-account` and `packages/interbit-template` applications. 
+This is one part of getting a working authentication loop happening 
```
working chain-authorization loop.

Commit:
Fix typo for interbit keys command

nicholeous commented on examples/platform-deploy.md:
```diff
@@ -0,0 +1,60 @@
+values from the GitHub OAuth app you just created. 
+1. The `PUBLIC_KEY` and `PRIVATE_KEY` values make up the Interbit key pair for 
+`platform-deploy`. A new key pair can be generated with the 
+`interbit key --filename my-platform-keys.json` command.  
```
interbit *keys* --filename meow.json

Perhaps also suggest making the keys in secrets/meow.json so nobody accidentally commits their secret key.

nicholeous commented on examples/platform-deploy.md:
````diff
@@ -0,0 +1,60 @@
+1. Open a terminal shell and build the Interbit modules
+```
+npm i
+npm run build:modules
````
This now runs automagically on postinstall

nicholeous commented on examples/platform-deploy.md:
````diff
@@ -0,0 +1,60 @@
+npm i
+npm run build:modules
+```
+1. Source the secrets file and generate the Interbit manifest file
````
To use their newly generated keys they will need to delete the old manifest file. This is because if there is a pre-existing manifest the genesis blocks will not be overwritten. At the moment there is a manifest with our existing pubkeys already in it.

The key that was generated will not work with the existing manfiest as our keys are specified in the genesisBlock ACL.

This will also have some side effects. When new genesis blocks are created this will also generate new chain IDs, as the chain ID is a hash of the genesis block.

Because of the new chain IDs, chain IDs throughout the monorepo will also be updated.

We kind of need to ask ourselves, who is this documentation for? If it is for internal use as well as developer use we need to mention the differences. 

For our own uses we need to acquire a valid key pair from someone and use it in the secrets file. We will NOT delete the manifest file.

nicholeous commented on examples/platform-deploy.md:
````diff
@@ -0,0 +1,60 @@
+```
+. platform-deploy.sh
+cd packages/platform-deploy
+npm run build:dev
````
Deploy runs the prod build so build:dev is not necessary.

nicholeous commented on examples/platform-deploy.md:
````diff
@@ -0,0 +1,60 @@
+cd packages/platform-deploy
+npm run build:dev
+```
+1. Deploy the platform with `npm start`
````
Having some commands in a shell section and some as 1. items in a list is confusing.

nicholeous replied:
This file should be separated into init - things that need to be configured once (ie. secrets setup) and starting the node (sourcing secrets and running start)

nicholeous commented on examples/web-auth-endpoint.md:
```diff
@@ -0,0 +1,58 @@
+
+### Set up a secrets file with GitHub tokens and node keys
+
+1. Create a `/secrets/web-auth-endpoint.sh` file with the following content. 
```
Please mention the cwd

nicholeous commented on examples/web-auth-endpoint.md:
````diff
@@ -0,0 +1,58 @@
+```
+1. The `GITHUB_CLIENT_ID` and `GITHUB_CLIENT_SECRET` are the same as in your 
+`secrets/platform-deploy.sh` file.
+1. The `PUBLIC_KEY` and `PRIVATE_KEY` values make up the Interbit key pair for 
````
These keys ought to have been generated during platform deploy setup. Because they are being generated now, this will change the keys in the genesis block ACL to include the web-auth keys, which will change the chain IDs, etc.

nicholeous commented on examples/web-auth-endpoint.md:
```diff
@@ -0,0 +1,58 @@
+`secrets/platform-deploy.sh` file.
+1. The `PUBLIC_KEY` and `PRIVATE_KEY` values make up the Interbit key pair for 
+`web-auth-endpoint`. A new key pair can be generated with the 
+`interbit key --filename my-web-auth-keys.json` command.  
```
interbit *keys* --filename meow.json

Commit:
WIP: updates to the platform-deploy doc

Commit:
Break up platform deploy docs into separate files

Commit:
Fix broken links and typos

karikushi commented:
@nicholeous I addressed your comments and restructured these docs. They're now framed as an example of a simple Interbit platform, which should (hopefully) give readers more context about what's going on with `platform-deploy` and `web-auth-endpoint`. I hit a snag with getting the c-auth loop working from the template app so that part is still pending (at the end of the user walk-through).

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+
+Our platform has two server nodes. One runs the `platform-deploy` hypervisor 
+and the other runs the `web-auth-endpoint` hypervisor. We also have two 
+browser nodes - one for the accounts app and another for the template app. 
```
I feel like this sentence contains an implied inversion of control. We have two apps which load n browser nodes - the nodes do not load the apps. The browser nodes are transient and only run for as long as a user has the app open in their browser though this sentence implies the nodes are always there.

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+Our platform has two server nodes. One runs the `platform-deploy` hypervisor 
+and the other runs the `web-auth-endpoint` hypervisor. We also have two 
+browser nodes - one for the accounts app and another for the template app. 
+These nodes run the React apps in the browser.
```
These react apps start an Interbit node in each browser that loads them.

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+These nodes run the React apps in the browser.
+
+The `platform-deploy` node hosts the accounts and template chains, acting as 
+the server for the two apps. It is the blockmaster and thus is responsible for 
```
I think we should shy away from the term server. This is not actually a server for the apps or a server at all - it is an Interbit node. The apps are served from their own static servers elsewhere in our production environment, and in dev mode they are served from their own webpack servers.

djlmills replied:
I think the key concept is the difference in node lifecycles. The browser nodes are transient, but the back-end nodes are long running, keeping the chains alive for as long as they run (as near as we currently get to chain persistence).

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+for the accounts and template chains to be hosted on the same node. They could 
+be hosted on separate nodes and moreover each app's chains could be hosted on 
+multiple nodes for redundancy. (FIXME: Should we mention that this has not 
+been implemented yet?) 
```
I think it's fine not to mention it. We're just stating that it's possible, not that we're doing it in practice.

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+multiple nodes for redundancy. (FIXME: Should we mention that this has not 
+been implemented yet?) 
+
+The `web-auth-endpoint` node hosts the accounts app chains, which includes a 
```
Great use of node and chains in context. Instead of using the term "which includes" consider using "in addition to"

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+GitHub chain. It also runs an Express instance that fields the http requests 
+coming from GitHub during an OAuth loop. 
+
+These two server nodes are peers. During the GitHub OAuth loop, OAuth requests 
```
~server nodes~ nodes

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+
+These two server nodes are peers. During the GitHub OAuth loop, OAuth requests 
+from  GitHub are received by the Express instance. An action is dispatched to 
+the GitHub chain on the `web-auth-endpoint` hypervisor which then passes it 
```
~on the~ into the

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+These two server nodes are peers. During the GitHub OAuth loop, OAuth requests 
+from  GitHub are received by the Express instance. An action is dispatched to 
+the GitHub chain on the `web-auth-endpoint` hypervisor which then passes it 
+along to the `platform-deploy` hypervisor. The `app-account` control chain 
```
While the idea of passing the action along to the other hypervisor is correct it ignores several steps and makes it difficult to understand the underlying behaviour. I think mentioning the chain joining and gossiping that makes this action passing possible would help clarify. Alternatively, finding different wording that is more vague but still descriptive enough to indicate gossip/joining would work too.

djlmills replied:
I think there is some confusion between cli and hypervisor. The express instance uses the web-auth-endpoint cli to dispatch an action that is distributed to, and processed deterministically by all nodes on the network. The chains and inter-node connections are managed by the hypervisor.

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+the GitHub chain on the `web-auth-endpoint` hypervisor which then passes it 
+along to the `platform-deploy` hypervisor. The `app-account` control chain 
+sees the new action and makes a new block on the user's private account chain. 
+This new block is then propagated throughout the network and is updated on the 
```
This sentence is a great example of wording that keeps the action propagation/gossiping/joining idea clear enough not to obscure, but vague enough not to distract from what you're describing.

nicholeous commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+`web-auth-endpoint` node. 
+
+When the `platform-deploy` package is built, it:
+- generates a config file by merging the config files at 
```
~generates a~ uses the -- a config file is always created by a user. Manifest files are automatically generated.

nicholeous commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+
+Our package script runs a post-install step which builds all the modules in 
+the Interbit repo. The build step for `platform-deploy` creates a new manifest 
+file with the newly generated keys and hashed covenant files. The build step 
```
Consider: ~with the~ using the

btlewan replied:
Hyphenate compound adjectives: `...the newly-generated keys...`

nicholeous commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+also updates the chain IDs for the other modules, including `app-account` and 
+`interbit-template`. 
+
+The `start` command spins up the `platform-deploy` server node, which listens 
```
~server~ node

nicholeous commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+`interbit-template`. 
+
+The `start` command spins up the `platform-deploy` server node, which listens 
+for incoming connections from the `web-auth-endpoint` server node and the 
```
~server~ node

nicholeous commented on examples/simple-platform/run-example.md:
````diff
@@ -0,0 +1,144 @@
+npm start
+```
+
+We've now started the browser nodes for the accounts and template apps. Your 
````
~browser nodes~ webpack dev servers which will serve our react apps. (which will start browser nodes when they are opened in a browser)

nicholeous commented on examples/simple-platform/run-example.md:
````diff
@@ -0,0 +1,144 @@
+```
+
+We've now started the browser nodes for the accounts and template apps. Your 
+browser should open a tab at `http://localhost:3025` with the accounts app and 
````
When the browser tab opens, it will boot an Interbit node to connect to the peer nodes running in the other terminal windows.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,19 @@
+# A Simple Interbit Platform 
+
+This example demonstrates how one might create a simple Interbit platform. Our
```
We don't include a definition of platform here, and we're using `platform` in a specific way. Let's add the following clause to the end of the first sentence:

```
...simple Interbit platform, which is a combination of individual Interbit apps.
```

btlewan replied:
Also, we haven't stated what problem we're trying to solve; that should be up-front. Rather than starting with what the platform is (or is going to be), state what it is for. Then describe how that would be accomplished.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,19 @@
+
+This example demonstrates how one might create a simple Interbit platform. Our
+platform has an accounts app (`packages/app-account`) which manages user 
+identities, and a second app (`packages/interbit-template`) that connects with 
```
When we say "connects with" to describe the template app, is that what the app does now, or is that what you get when the platform app is created? If it's the former, why would we need to create the platform? If it's the latter, the ambiguity needs to be removed: just say what the template app does.

Or, consider rewriting the sentence like this:

We'll create our platform by combining two apps, `packages/app-account` (which manages user identities) and `packages/interbit-template` (which creates a single blockchain). The resulting platform app lets us restrict blockchain updates to authenticated users.

btlewan commented on examples/simple-platform/platform-structure.md:
```diff
@@ -0,0 +1,45 @@
+# Platform structure
```
Consider renaming this file to just `structure.md`. The title itself is fine, but the filename implies that there might be other kinds of structure topics in this folder.

karikushi replied:
👍

btlewan commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+# Running the Platform Example Locally
```
Consider renaming this file to just `run.md`. The `example` is a bit redundant (since it's in the `examples` folder), and could imply that there may be other `run` topics in this folder.

Also, the title includes the `example` redundancy, and it's better to avoid gerunds (verbs converted to nouns by adding `ing`): `Run the Platform Locally`

karikushi replied:
👍

btlewan commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+
+### Create a GitHub development application for the `account-app` OAuth process
+
+At this time, the accounts app supports account creation and authentication 
```
Convert the first sentence into an admonition:

```
> At this time...
```

btlewan commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+1. Log into [github.com](https://github.com) and navigate to Settings > 
+Developer settings. 
+1. Under OAuth Apps, create a new OAuth App and give it any application name, 
+e.g. Interbit Test App.
```
Ideally, user input should appear in monospace, to delineate the input from the discussion. So:

```
... e.g. `Interbit Test App`.
```

btlewan commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+1. Under OAuth Apps, create a new OAuth App and give it any application name, 
+e.g. Interbit Test App.
+1. The Authorization callback URL is `http://localhost:8888/oauth/github`.
+1. Click the Register Application button. Leave this browser tab open as we 
```
It's helpful to identify labelled UI elements with boldface. So:

```
1. Click the **Register Application** button. Leave...
```

btlewan commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+Developer settings. 
+1. Under OAuth Apps, create a new OAuth App and give it any application name, 
+e.g. Interbit Test App.
+1. The Authorization callback URL is `http://localhost:8888/oauth/github`.
```
This is a declaration, not an imperative instruction. Procedures that have sub-steps should be composed like this:

```
1.  **Imperative instruction**
    1.  Sub-step.
        Sub-step's discussion (if required)
    1.  Sub-step.
    ...
    Discussion of overall step (if required)

1.  **Imperative instruction**
    Discussion of instruction (if required)
...
```

Providing this structure allows users to better identify where their progress, and lets them skim the high-level instructions faster.

djlmills replied:
Some points that should be conveyed here:
The callback URL is case sensitive.
`localhost` is only for development. 
The port number is that specified for running the `web-auth-endpoint` node locally.

btlewan commented on examples/simple-platform/run-example.md:
````diff
@@ -0,0 +1,144 @@
+interbit keys --filename platform-deploy-keys.json
+interbit keys --filename web-auth-endpoint-keys.json
+```
+This will create two new files at the root of the Interbit repository with 
````
Try to avoid future tense wherever possible, because it implies a promise to the reader that we might not be able to fulfill. In this case, it also creates uncertainty: when should I expect the new files to be created?

The pattern to follow is "do this then that happens".

So: `This creates two new files...`

btlewan commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+### Set up environment variables for `platform deploy`
+
+1. Create a `/secrets` directory at the root of the Interbit repository. This 
+directory is ignored by .git. 
```
No period in front of `git`.

Regular users of git know about `.gitignore`. New users might not know, and might assume that this sentence is telling that `/secrets` is ignored always. If it's important to point out that the folder is ignored, it's important to say how/why.

btlewan commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+
+1. Create a `/secrets` directory at the root of the Interbit repository. This 
+directory is ignored by .git. 
+1. Create a `/secrets/platform-deploy.sh` file with the following content
```
I find it helpful when editing topics to include a blank line between steps, especially if each step occupies multiple source lines.

btlewan commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+1. Copy the `GITHUB_CLIENT_ID`, `GITHUB_CLIENT_SECRET`, and 
+`GITHUB_REDIRECT_URL` 
+values from the GitHub OAuth app you just created. 
+1. The `PUBLIC_KEY` and `PRIVATE_KEY` values come from 
```
This is a declaration, not an imperative instruction.

btlewan commented on examples/simple-platform/run-example.md:
````diff
@@ -0,0 +1,144 @@
+export PORT=8888
+export CONNECT_TO_PEERS="localhost:5025"
+```
+1. The `GITHUB_CLIENT_ID` and `GITHUB_CLIENT_SECRET` are the same as in your 
````
This step and the next one are declarations, not imperative instructions.

btlewan commented on examples/simple-platform/run-example.md:
```diff
@@ -0,0 +1,144 @@
+
+### Build and start the `platform-deploy` node
+
+1. Delete the existing manifest file, 
```
Split this paragraph after the first sentence, so that you have an imperative instruction for this procedure step, and a paragraph of discussion.

btlewan commented on examples/simple-platform/run-example.md:
````diff
@@ -0,0 +1,144 @@
+repo's root:
+```
+npm i
+. platform-deploy.sh
````
Replace the initial period with `source`. Not everyone uses bash as their shell, and not all shells include the dot alias.

djlmills replied:
Yup. `source` is needed for Windows.

btlewan commented on examples/simple-platform/run-example.md:
````diff
@@ -0,0 +1,144 @@
+npm start
+```
+
+Our package script runs a post-install step which builds all the modules in 
````
`...builds all of the modules...`

btlewan reviewed (changes requested):
This is looking much better! A number of my comments affect multiple instances within the PR, so I'll let you tackle those first before adding additional comments.

djlmills commented on examples/simple-platform/run-example.md:
````diff
@@ -0,0 +1,144 @@
+Open a new terminal shell and run the following commands from the `interbit` 
+repo's root: 
+```
+. /secrets/web-auth-endpoint.sh
````
`source` again.

djlmills commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,48 @@
+Open the accounts app in your browser at `http://localhost:3025`. If you open 
+your browser's developer tools to the Console tab, you can see some 
+`interbit-middleware` commands being printed. The accounts app is fully up and 
+running once you see the `interbit-middleware/CHAIN_BLOCKING` status for the 
```
These are not commands. `log messages`.

djlmills commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,48 @@
+`interbit-middleware` commands being printed. The accounts app is fully up and 
+running once you see the `interbit-middleware/CHAIN_BLOCKING` status for the 
+`accountsPrivate` chain. This may take up to a minute. 
+
```
It really shouldn't take up to a minute. It does take longer the first time.

It is also useful to clarify what you mean by fully up and running. What you mean is that you can dispatch actions to your private account chain.

djlmills commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,48 @@
+prompted to enter your GitHub credentials. Enter your credentials and sign in. 
+You should be redirected back to the accounts app. 
+
+You will see a number of chain actions being dispatched in the console. 
```
Which console? It is generally helpful to monitor the `platform-deploy` console to see actions that are dispatched to the account and template chains.

Commit:
Update platform structure and run example docs based on PR feedback

Commit:
Update intro to platform example docs

Commit:
Rename files in platform example

djlmills commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,147 @@
+At this time, the accounts app supports account creation and authentication 
+via GitHub only. In order to run a user authentication loop, you need to 
+create a new GitHub OAuth app and configure it to connect to the 
+`web-auth-endpoint` node.
```
redirect to the Express server on the `web-auth-endpoint` node.

Commit:
Fix typos and line lengths

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+of apps: Interbit platform apps and apps built on the platform. 
+
+By platform apps, we mean apps intended to make Interbit app development 
+easier by solving common problems such as user identity management, payment 
```
Add a comma before `such`: `...common problems, such as...`. Why? Because the entire `such as clause` is optional, or _non-restrictive_.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+By platform apps, we mean apps intended to make Interbit app development 
+easier by solving common problems such as user identity management, payment 
+processing, and hosting. These apps make up the Interbit platform. They may 
+not provide much value to the user on their own - after all, how useful is a 
```
Change the single dash into an em dash (`&mdash;`). An em dash can serve the purpose of a comma, colon, or parentheses. A single, or _en_ dash (which should be slightly wider than a hyphen, typographically) should only be used to represent "to" when describing a range, for example April-May, or 2-5.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+not provide much value to the user on their own - after all, how useful is a 
+payment processing app if there is nothing to pay for? However, they allow 
+developers to concentrate their efforts on solving app-specific problems 
+instead of reinventing the wheel. Apps of the second type of are those built 
```
The bulk of the paragraph describes platform apps. The description of built-on-the-platform apps should become its own paragraph.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+instead of reinventing the wheel. Apps of the second type of are those built 
+by developers that use the functionality provided by the platform.
+
+This example has one platform app, `packages/app-account`, which provides user 
```
The use of `This` seems ambiguous, given the context in the preceding paragraph. Also, I'm not keen on `has`, because the example doesn't own or possess the platform app. Could we instead say:

```
The following example includes a platform app called the Accounts app,
`packages/app-account` -- providing user authentication and identity
management functionality -- and a Template app, `packages/interbit-template`
-- a simple starting point for creating your own Interbit apps.
```

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+applications. 
+
+The goal of this example is to demonstrate how to request and share user 
+profile information from outside the accounts app. We configure the platform 
```
There is some redundancy between this paragraph, starting with `We configure...`, and the following bulleted list. I'd suggest keeping the first sentence, and replace `We configure...`, to the end of the paragraph, with:

```
To achieve that goal, we cover:
```

And convert the `We configure...` and `In order to...` sentences into bullet points.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+chain from the template app. In order to demonstrate this, we also create a 
+user account that is authenticated via GitHub. Specifically, we go over:
+
+- the platform structure: configuration of the apps' chain joins and peers
```
Convert each bullet point into full sentences.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+- chain hosting for our apps and the GitHub OAuth loop with 
+`packages/platform-deploy` and `packages/web-auth-endpoint`
+- how to run the example locally
+- a walk-through of the accounts and templates apps' user interface to create 
```
Since there are two apps, there are two user interfaces, so `interface` should be plural.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+`packages/platform-deploy` and `packages/web-auth-endpoint`
+- how to run the example locally
+- a walk-through of the accounts and templates apps' user interface to create 
+an account and authorize profile information sharing
```
Comma before `and`: `...an account, and authorize...`.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+- a walk-through of the accounts and templates apps' user interface to create 
+an account and authorize profile information sharing
+
+The example is divided into the following documents:
```
We call them `topics`, not `documents`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+# Running the Platform Example Locally
```
Update the title to match the link label in `README.md`: `Run the Example Locally`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+## Setup
+
+### Create a GitHub development application for the `account-app` OAuth process
```
`account-app` is incorrect. It's `app-account`. However, we don't need to use the technical package name here: `...Accounts app's OAuth process`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+### Create a GitHub development application for the `account-app` OAuth process
+
+The accounts app supports account creation and authentication via GitHub only. 
```
Capitalize `Accounts`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+### Create a GitHub development application for the `account-app` OAuth process
+
+The accounts app supports account creation and authentication via GitHub only. 
+In order to run a user authentication loop, you need to create a new GitHub 
```
`loop` is incorrect. The process of obtaining an OAuth token is called a `flow`. We should say:

```
In order to complete an OAuth flow with GutHub as the identity provider, you need...
```

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+In order to run a user authentication loop, you need to create a new GitHub 
+OAuth app and configure it to connect to the `web-auth-endpoint` node.
+
+1. Go to [github.com](https://github.com) and log into your account.
```
URLs that point to the top of a domain, or to a directory, should end in a slash `/` to avoid unnecessary redirects.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+1.  Click the **Register application** to finish creating the OAuth app.
+
+1. Leave this browser tab open as we will copy values from here in the steps 
```
Add a comma after `open`. Also avoid future tense:

```
Leave this browser tab open,  as we are going to copy values
from this tab in subsequent steps.
```

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+### Generate a set of private and public keys for `platform-deploy` and `web-auth-endpoint`
+
+1. Open a terminal and install `interbit-cli` with
```
Add a colon after `with`.

btlewan replied:
Also, add a blank line after the imperative and the example/discussion.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+```
+npm i -g interbit-cli
+```
+1. Generate an Interbit key pair for `platform-deploy` and 
````
Add a blank line between steps.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+npm i -g interbit-cli
+```
+1. Generate an Interbit key pair for `platform-deploy` and 
+`web-auth-endpoint`, respectively, with
````
Add a colon after `with`, and add a blank line after this line.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+interbit keys --filename web-auth-endpoint-keys.json
+```
+
+  This creates two new files at the root of the Interbit repository with your 
````
How does the `interbit` tool know where the repository is? If I execute these commands and my current directory is `/tmp`, is the statement still true? If not, we need a change-directory step that precedes this one.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+### Set up environment variables for `platform deploy`
+
+1. Create a `/secrets` directory at the root of the Interbit repository. 
```
Do we literally require a directory name that starts with slash? That might be quite difficult (or impossible) to accomplish on some filesystems. If you mean `<Interbit repository root>/secrets`, just say `secrets`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+1. Create a `/secrets` directory at the root of the Interbit repository. 
+
+   This directory is included in this repository's gitignore file and will not 
```
The filename is `.gitignore`, and filenames should appear in monospace. And avoid future tense. Finally, why is git tracking (or lack thereof) important?

```
...file, and is not get tracked by git. That means that the secrets contained
within cannot be accidentally committed to the repo.
```

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+   This directory is included in this repository's gitignore file and will not 
+be tracked by git.
+
+1. Create a `/secrets/platform-deploy.sh` file with the following content
```
````
1. Change the current working directory to the `secrets` directory:
   ```sh
   cd secrets
   ```

1. Create the file `platform-deploy.sh` containing the following content:
````

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+export CONNECT_TO_PEERS="localhost:8888"
+```
+
+1. Copy the `GITHUB_CLIENT_ID`, `GITHUB_CLIENT_SECRET`, and 
````
```
1. Copy the values for the `GITHUB_CLIENT_ID`, `GITHUB_CLIENT_SECRET`, and
   `GITHUB_REDIRECT_URL` fields from the GitHub OAuth app you just created
   (in the open browser tab), and paste them into the respective fields in
   the `platform-deploy.sh` file.
```

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+1. Copy the `GITHUB_CLIENT_ID`, `GITHUB_CLIENT_SECRET`, and 
+`GITHUB_REDIRECT_URL` values from the GitHub OAuth app you just created. 
+
+1. Copy `PUBLIC_KEY` and `PRIVATE_KEY` values from `platform-deploy-keys.json`.
```
```
1. Copy the values for the `PUBLIC_KEY` and `PRIVATE_KEY` fields from
   the `platform-deploy-keys.json` file, and paste them into the
   respective fields in the `platform-deploy.sh` file.
```

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+### Set up environment variables for `web-auth-endpoint`
+
+1. Create a `/secrets/web-auth-endpoint.sh` file with the following content. 
```
This section needs the same kind of alterations as in the previous section.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+`web-auth-endpoint-key.json`.
+
+
+## Spinning up the platform 
```
Maybe just `Start the platform`? Some readers might envision that rotation is somehow required.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+    Our package script runs a post-install step which builds all of the 
+modules in the Interbit repo. The build step for `platform-deploy` creates a 
+new manifest file using the newly generated keys and hashed covenant files. 
```
Hyphenate `newly-generated`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+generate a new manifest, we create new genesis blocks which results in new 
+chain IDs (which are hashes of the genesis blocks). 
+
+1. Open a terminal shell and run the following commands from the `interbit`
```
Omit `shell`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+incoming connections from the `web-auth-endpoint` node and the 
+`app-account` and `interbit-template` browser nodes. 
+
+1. Leave this terminal shell open. 
```
This doesn't need to be a step, because the desired effect is an absence of action. Just make it a sentence at the end of the discussion above. Also, omit `shell`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+### Start the `web-auth-endpoint` hypervisor on a node
+
+1. Open a new terminal shell and run the following commands from the `interbit` 
```
Omit `shell`.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+1. Open a new terminal shell and run the following commands from the `interbit` 
+repo's root: 
+```
+source /secrets/web-auth-endpoint.sh
````
No leading slash before `secrets`.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+npm start
+```
+
+1. Again, leave this shell open. 
````
This is a new terminal, so we can't leave it open again. It should be part of the previous discussion, not a step on its own. Replace`shell` with terminal.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+1. Leave this terminal shell open. 
+
+
+### Start the `web-auth-endpoint` hypervisor on a node
```
`...Web Auth Endpoint app's hypervisor...`

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+1. Again, leave this shell open. 
+
+
+### Start the app-account and interbit-template apps
```
`...the Account and Template apps`

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+### Start the app-account and interbit-template apps
+
+1. In a third terminal, run the following from the `interbit` repo root
```
```
1. In a new terminal (the third), run the following commands from the `interbit` repo root:
```

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+cd packages/app-account
+npm start
+```
+
````
Add `Leave this terminal open.`

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+npm start
+```
+
+1. And, in a fourth terminal 
````
```
1. In a new terminal (the fourth), run the following commands from the `interbit` repo root:
```

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+npm start
+```
+
+We've now started the webpack development servers which serve our React apps. 
````
Try to avoid unnecessary contractions: `We have...`

btlewan replied:
Is is correct to call these `webpack` servers? I thought webpack bundles assets/dependencies, but I was unaware that it was also a server.

karikushi replied:
Yes, in development mode they are servers. Webpack has hot loading capabilities so when app code is changed, it serves up the new version of the app without going through a manual compilation step or having to reload in the browser.

Commit:
Finish walk-through docs for the template app cAuth loop

karikushi commented:
@nicholeous and @btlewan Please have another read through these docs. I've reworked the intro section to frame the example and added the walk through of the template app c-auth loop.

nicholeous commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+payment processing app if there is nothing to pay for? However, they allow 
+developers to concentrate their efforts on solving app-specific problems 
+instead of reinventing the wheel. Apps of the second type of are those built 
+by developers that use the functionality provided by the platform.
```
Bam! Nice

nicholeous commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,106 @@
+
+## Allowing the template app to connect to your private account chain
+
+Now that we've created a user account on the accounts app, we'll authorize the 
```
+1 for precise use of authentication and authorisation.  Yesssss.

nicholeous reviewed (approved):
:smile_cat:

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,177 @@
+```
+
+We've now started the webpack development servers which serve our React apps. 
+Your browser should open a tab at `http://localhost:3025` with the accounts 
````
Capitalize `Accounts`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+
+We've now started the webpack development servers which serve our React apps. 
+Your browser should open a tab at `http://localhost:3025` with the accounts 
+app and at `http://localhost:3000` with the template app. When these tabs
```
Capitalize `Template`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+We've now started the webpack development servers which serve our React apps. 
+Your browser should open a tab at `http://localhost:3025` with the accounts 
+app and at `http://localhost:3000` with the template app. When these tabs
+open, the React apps will boot Interbit nodes to connect to the 
```
Omit `will`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+Your browser should open a tab at `http://localhost:3025` with the accounts 
+app and at `http://localhost:3000` with the template app. When these tabs
+open, the React apps will boot Interbit nodes to connect to the 
+`platform-deploy` and `web-auth-endpoint` peer nodes running in the other 
```
Omit `other`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+app and at `http://localhost:3000` with the template app. When these tabs
+open, the React apps will boot Interbit nodes to connect to the 
+`platform-deploy` and `web-auth-endpoint` peer nodes running in the other 
+terminal windows. 
```
We don't know that they are windows, so just end the sentence with `terminals.`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+`platform-deploy` and `web-auth-endpoint` peer nodes running in the other 
+terminal windows. 
+
+In the [next section](user-walk-through.md),  we go over the user flow for 
```
A very minor nit, but there are two spaces before `we`. Also, your editor likes to leave trailing spaces at the end of most lines.

karikushi replied:
I'll look into updating my editor's preferences. Our other repos have linters set up to remove trailing/extra white spaces.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+`platform-deploy` and `web-auth-endpoint` peer nodes running in the other 
+terminal windows. 
+
+In the [next section](user-walk-through.md),  we go over the user flow for 
```
Use `describe` instead of `go over`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+# Platform structure
+
+Our example has two React applications: `app-account` and 
+`interbit-template`. These apps are served in the user's browser as any other 
```
Add a comma after `browser`.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,39 @@
+
+- the platform structure: configuration of the apps' chain joins and peers
+- chain hosting for our apps and the GitHub OAuth loop with 
+`packages/platform-deploy` and `packages/web-auth-endpoint`
```
What are these? This is the first mention, and it's not clear if they are apps or something else. In `structure.md`, we call them hypervisors. We should do that here too, along with a mention of what a hypervisor is.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,177 @@
+below. 
+
+
+### Generate a set of private and public keys for `platform-deploy` and `web-auth-endpoint`
```
What are these things, `platform-deploy` and `web-auth-endpoint`? In `structure.md` we call them hypervisors. We should do so here too.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+`web-auth-endpoint` hypervisors. 
+
+The `platform-deploy` node hosts the accounts and template chains. (See 
+[Chain Sponsorship](../../reference/interbit-middleware/chainSponsorship.adoc) 
```
Remove the `../..` before `/reference`. GitBook knows where the top-level folder containing the documentation lives, and using an absolute path makes it easier if we need to move this example up or down a level.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+
+The `platform-deploy` node hosts the accounts and template chains. (See 
+[Chain Sponsorship](../../reference/interbit-middleware/chainSponsorship.adoc) 
+for more information on hosting chains.)  It is the blockmaster and thus is 
```
`blockmaster` should be two words.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+The `platform-deploy` node hosts the accounts and template chains. (See 
+[Chain Sponsorship](../../reference/interbit-middleware/chainSponsorship.adoc) 
+for more information on hosting chains.)  It is the blockmaster and thus is 
+responsible for adding new blocks to the application chains. Note that there 
```
Either remove the `Note that` and start the sentence with `There...`, or convert the sentences into an actual admonition.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+for more information on hosting chains.)  It is the blockmaster and thus is 
+responsible for adding new blocks to the application chains. Note that there 
+is no requirement for the accounts and template chains to be hosted on the 
+same node. They could be hosted on separate nodes and moreover each app's 
```
Omit `moreover`. If you feel that the emphasis is required: `...nodes and, more importantly, each app's...`

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+chains could be hosted on multiple nodes for redundancy. 
+
+The `web-auth-endpoint` node hosts the accounts app chains and a GitHub chain. 
+It also runs an Express instance that fields the http requests coming from 
```
Capitalize `HTTP`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+
+The `web-auth-endpoint` node hosts the accounts app chains and a GitHub chain. 
+It also runs an Express instance that fields the http requests coming from 
+GitHub during an OAuth loop. 
```
Use `flow` instead of `loop`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+It also runs an Express instance that fields the http requests coming from 
+GitHub during an OAuth loop. 
+
+These two nodes are peers. During the GitHub OAuth loop, OAuth requests 
```
`flow` instead of `loop`. GitHub doesn't own OAuth or the flow, they're just an identity provider, so:

```
During the OAuth flow, OAuth requests from Github...
```

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+`packages/interbit-template/interbit.config.js`. These files specify peer 
+nodes, static chains, and chain join configurations. See the [Template App 
+Walk-through](../template.adoc) for a detailed explanation of the template 
+app's chain architecture 
```
Add a period after `architecture`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+- generates a manifest file at 
+`packages/platform-deploy/platform/interbit.manifest.json`. See 
+[Interbit CLI Manifest](../../reference/interbit-cli/manifest.adoc).
+- creates hashed covenant files for the accounts and template app covenants. 
```
Capitalize `Creates`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+nodes, static chains, and chain join configurations. See the [Template App 
+Walk-through](../template.adoc) for a detailed explanation of the template 
+app's chain architecture 
+- generates a manifest file at 
```
Capitalize `Generates`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+managed by both hypervisors. 
+
+When the `platform-deploy` package is built, it:
+- uses a config file that is a union of the config files at 
```
Capitalize `Uses`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+The `web-auth-endpoint` config file is the same as the `app-account` config 
+file. 
+
+Both apps and their chains specify `platform-deploy` as a peer node. The 
```
Do we want to say this?

```
...specify the `platform-deploy` hypervisor as a peer node.
```

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,47 @@
+file. 
+
+Both apps and their chains specify `platform-deploy` as a peer node. The 
+accounts app and its chains also specify `web-auth-endpoint` as a peer. 
```
Capitalize `Accounts`.

Do we want to say?

```
...specify the `web-auth-endpoint` hypervisor as a peer.
```

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+# User Walk-through of the Platform Example
+
+## Creating an account and going through the GitHub OAuth loop
```
`Create an account and complete the OAuth flow with GitHub`

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+your browser's developer tools to the console tab, you can see some 
+`interbit-middleware` log messages being printed. The accounts app is fully up 
+and running once you see the `interbit-middleware/CHAIN_BLOCKING` status for 
+the `accountsPrivate` chain. At this point, you will be able to dispatch 
```
Avoid future tense: `At this point, you can dispatch...`

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+actions to your private account chain.  
+
+From the main accounts app page, click on the **Create Account / Sign in** 
+link in the header, which navigates you to the `create-account` page. Click 
```
Replace `navigates you to` with `displays`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+From the main accounts app page, click on the **Create Account / Sign in** 
+link in the header, which navigates you to the `create-account` page. Click 
+the **Create Account** button, check that you have read the disclaimer and 
+click **Continue** in the modal.
```
`modal dialog`, or just `dialog`, would be better than `modal`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+the **Create Account** button, check that you have read the disclaimer and 
+click **Continue** in the modal.
+
+If you're already signed into your GitHub account, you will be directed to the 
```
Avoid future tense:

```
If you're already signed into your GitHub account, your personal account page is displayed.
```

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+click **Continue** in the modal.
+
+If you're already signed into your GitHub account, you will be directed to the 
+your personal account page. Otherwise, you will be redirected to GitHub and 
```
Avoid future tense: `Otherwise, the GitHub login dialog appears.`

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+prompted to enter your GitHub credentials. Enter your credentials and sign in. 
+You should be redirected back to the accounts app. 
+
+You will see a number of chain actions being dispatched in the browser console 
```
Avoid future tense: `You should see...`

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+on the accounts app tab. Specifically, after the action type 
+`app-account-my-account/COMPLETE_AUTHENTICATION` is dispatched, the 
+`chainState.profile` property is updated with your GitHub identity 
+information. You can also verify that the OAuth loop was successful if you see 
```
Replace `loop` with `flow`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+information. You can also verify that the OAuth loop was successful if you see 
+your GitHub username appear in the header. 
+
+Congratulations, you have now created an Interbit account. You can make 
```
`Congratulations! You have now...`

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+your GitHub username appear in the header. 
+
+Congratulations, you have now created an Interbit account. You can make 
+changes to your personal account information and see how the chain blocks with 
```
What is meant by `blocks`? It's terminology common to experienced blockchain developers. Readers of this example need to know that you mean "new blocks added to the blockchain", as opposed to "prevents other actions from taking place".

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+every change. The **Block Explorer** page has a handy user interface for 
+viewing the blocks and the private account chain's state. 
+
+## Allowing another application to connect to your private account chain
```
Replace `Allowing` with `Allow`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+
+## Allowing another application to connect to your private account chain
+
+Now that we've created a user account on the accounts app, we'll authorize the 
```
Avoid unnecessary contractions: `we have` and `we can`

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+
+Open the template app in your browser at `http://localhost:3000`. You should 
+see the browser console print out `interbit-middleware/CHAIN_BLOCKING` for the 
+`templatePrivate` chain. You'll also see a printout of the chain state on the 
```
Avoid unnecessary contractions: `You should also see...`

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+`templatePrivate` chain. You'll also see a printout of the chain state on the 
+template app's `/chains` page. 
+
+We'll now make a chain authorization request from the template app to the 
```
Avoid future tense and unnecessary contractions: `We can now make...`

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,49 @@
+
+We'll now make a chain authorization request from the template app to the 
+account app. Click the **Connect to Profile** link in the header to navigate to 
+the `/cauth/request` page and click on the **Request cAuth** button. You will 
```
Avoid future tense; replace `You will be redirected to the accounts app.` with `The Accounts app is displayed.`

btlewan reviewed (changes requested):
Lots of mostly minor fixes. I really like the new intro!

btlewan commented:
It looks like a bunch of my comments are marked as outdated because you updated the PR while I was reviewing. That's probably the weakest part of GitHub's review process.

Commit:
Make revisions based on PR feedback

Commit:
Fix a typo and line length error

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,47 @@
+
+Before we dive into our Interbit platform example, let's take a step back and
+define the ecosystem of Interbit applications. We are concerned with two types
+of apps: Interbit platform apps and apps built on the platform.
```
Add a comma after the second instance of `apps`: `...platform apps, and apps built...`

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,47 @@
+cover:
+
+- How to configure the platform and Template app so that a user can grant read
+access to their private accounts chain from the Template app
```
Add a period at the end of the sentence.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,47 @@
+- How to configure the platform and Template app so that a user can grant read
+access to their private accounts chain from the Template app
+- How to create a user account that is authenticated via GitHub so that the
+user has a private accounts chain to share
```
Add a period at the end of the sentence.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,47 @@
+
+- Give an overview of the platform structure by:
+  - describing how the apps' chain joins and peers are configured
+  - describing the chain hosting for our apps and the GitHub OAuth loop with
```
Replace `loop` with `flow`.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,47 @@
+Specifically, we:
+
+- Give an overview of the platform structure by:
+  - describing how the apps' chain joins and peers are configured
```
Add a comma at the end of the line, after `configured`.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,47 @@
+- Give an overview of the platform structure by:
+  - describing how the apps' chain joins and peers are configured
+  - describing the chain hosting for our apps and the GitHub OAuth loop with
+the `packages/platform-deploy` and `packages/web-auth-endpoint` hypervisors
```
Add a period after `hypervisors`.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,47 @@
+  - describing how the apps' chain joins and peers are configured
+  - describing the chain hosting for our apps and the GitHub OAuth loop with
+the `packages/platform-deploy` and `packages/web-auth-endpoint` hypervisors
+- Give instructions on how to run the example locally
```
Add a period at the end of the sentence.

btlewan commented on examples/simple-platform/README.md:
```diff
@@ -0,0 +1,47 @@
+the `packages/platform-deploy` and `packages/web-auth-endpoint` hypervisors
+- Give instructions on how to run the example locally
+- Give a walk-through of the accounts and templates apps' user interfaces to
+create an account, and authorize profile information sharing
```
Add a period at the end of the sentence.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+# Run the Platform Example Locally
```
`Run the Example locally`

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+
+## Setup
+
+### Create a GitHub development application for the Accounts app OAuth process
```
Replace `process` with `flow`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+1. Click the **New OAuth App** button to create a new OAuth app and configure
+it.
+
+1. In the **Application name** field, give the app a name of your choosing,
```
Replace `choosing` with `choice`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+
+1.  Click the **Register application** to finish creating the OAuth app.
+
+1. Leave this browser tab open as we are going to copy values from this tab in
```
Make this part of the previous list item; it's a non-action so it doesn't deserve to be a step on its own.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+
+### Generate a set of private and public keys for the `platform-deploy` and `web-auth-endpoint` hypervisors
+
+1. Open a terminal and install `interbit-cli` with:
```
Omit `with`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+  public and private key pairs.
+
+
+### Set up environment variables for `platform deploy`
```
Replace the space between `platform` and `deploy` with a hyphen: `platform-deploy`

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,205 @@
+    npm i -g interbit-cli
+    ```
+
+1. From the root of the `interbit` repository, generate an Interbit key pair
````
In this list item, we say:

```
...`interbit` repository, ...
```

In the next, we say:

```
...Interbit repository.
```

I think the latter is better, but I'm not too fussy. However, they should be consistent. Whichever change you make, make it throughout the example.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,205 @@
+    ```
+
+1. Copy the values for the `GITHUB_CLIENT_ID` and `GITHUB_CLIENT_SECRET` fields
+from the GitHub OAuth app you just created. These values are the same as in
````
After `created`, add: `(in the open browser tab)`. It's always good to be clear where values are coming from and going to. The next sentence is good, and reinforces the point, but doesn't describe the source/destination directly.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+
+   This file must be deleted when new Interbit key pairs are generated (as we
+did in our setup). If there is a pre-existing manifest, the genesis blocks
+will not be overwritten and our new keys will not work. Note that when we
```
Replace the first `will not` with `are not`. Replace the second `will not` with `do not`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+### Start the Accounts and Template apps
+
+1. In a new terminal (the third), run the following commands from the
+`interbit` repo root:
```
`repo` should be `repository`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+    Leave this terminal open.
+
+1. In a new terminal (the fourth), run the following commands from the
+`interbit` repo root:
```
`repo` should be `repository`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+### Start the Web Auth Endpoint hypervisor on a node
+
+1. Open a new terminal and run the following commands from the `interbit`
+repo's root:
```
`repo's` should be `repository's`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+chain IDs (which are hashes of the genesis blocks).
+
+1. Open a terminal and run the following commands from the `interbit`
+repo's root:
```
`repo's` should be `repository's`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+
+We have now started the webpack development servers which serve our React apps.
+Your browser should open a tab at `http://localhost:3025` with the Accounts
+app and at `http://localhost:3000` with the Template app. When these tabs
```
replace `app and at` with `app, and another tab at`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+We have now started the webpack development servers which serve our React apps.
+Your browser should open a tab at `http://localhost:3025` with the Accounts
+app and at `http://localhost:3000` with the Template app. When these tabs
+open, the React apps boot Interbit nodes to connect to the
```
Replace `boot` with `start`. A portion of our readers won't understand `boot`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,44 @@
+
+Our example has two React applications: `app-account` and `interbit-template`.
+These apps are served in the user's browser, as any other React app would be.
+Either there is a static bundle served in a production environment, or they
```
This sentence needs clarification, but I'm not entirely sure how to say it. Is there one static bundle for both apps? By `they`, are we still referring to the apps or to their assets+code?

Is the point you're making here relevant? Could we just drop the entire sentence?

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,44 @@
+Interbit node each for the `platform-deploy` and `web-auth-endpoint`
+hypervisors.
+
+The `platform-deploy` node hosts the accounts and template chains. (See
```
Capitalize `Accounts` and `Template`, unless you mean chains that are literally called `accounts` and `chains`, in which case, make them monospace.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,44 @@
+
+The `platform-deploy` node hosts the accounts and template chains. (See
+[Chain Sponsorship](/reference/interbit-middleware/chainSponsorship.adoc)
+for more information on hosting chains.)  It is the block master and thus is
```
Replace `It is` with `This node`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,44 @@
+[Chain Sponsorship](/reference/interbit-middleware/chainSponsorship.adoc)
+for more information on hosting chains.)  It is the block master and thus is
+responsible for adding new blocks to the application chains. There is no
+requirement for the accounts and template chains to be hosted on the same node.
```
Capitalize (or monospace, but not both) `accounts` and `template`, to be consistent with the earlier use in this paragraph.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,44 @@
+multiple nodes for redundancy.
+
+The `web-auth-endpoint` node hosts the Accounts app chains and a GitHub chain.
+It also runs an Express instance that fields the HTTP requests coming from
```
Replace `instance` with `web server`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,44 @@
+GitHub during an OAuth flow.
+
+These two nodes are peers. During the GitHub OAuth flow, OAuth requests from
+GitHub are received by the Express instance. The Express instance uses the
```
Replace both mentions of `instance` with `web server`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,44 @@
+app's chain architecture.
+- Generates a manifest file at
+`packages/platform-deploy/platform/interbit.manifest.json`. See
+[Interbit CLI Manifest](../../reference/interbit-cli/manifest.adoc).
```
Remove `../..` before `/reference`.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,44 @@
+`packages/platform-deploy/platform/interbit.manifest.json`. See
+[Interbit CLI Manifest](../../reference/interbit-cli/manifest.adoc).
+- Creates hashed covenant files for the Accounts and Template app covenants.
+See [Covenants](../../key-concepts/covenants.adoc).
```
Remove `../..` before `/key-concepts`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+# User Walk-through of the Platform Example
```
Links to this topic just say `User Walk-through`. I've suggested `Platform Walk-through` above. Pick one or the other, and make it consistent throughout.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+see the `interbit-middleware/CHAIN_BLOCKING` status logged to the console for
+the `accountsPrivate` chain.
+
+   The Interbit middleware prints a number of messages to the console so you
```
Insert `that` between `so` and `you`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+
+   The Interbit middleware prints a number of messages to the console so you
+can see, for instance, when chains have loaded and actions have been
+dispatched. The account app is fully up and running when you see the above
```
Replace `account` with `Accounts`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+1. Sign in to your GitHub account. If you're already logged into your GitHub
+account, your personal account page is displayed. Otherwise, the GitHub
+sign-in page appears. Enter your GitHub credentials and sign in to your
+account. You should then be redirected back to the accounts app's personal
```
Capitalize `Accounts`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+account page.
+
+   You should see a number of chain actions being dispatched in the browser
+console on the accounts app tab. Specifically, after the action type
```
Capitalize `Accounts`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+can see, for instance, when chains have loaded and actions have been
+dispatched. The account app is fully up and running when you see the above
+message. At this point, you can dispatch actions to your
+private account chain.
```
Is there one chain per Interbit account? If so, we should say `...to the private blockchain containing your account information`. If not, we should replace `account` with `accountsPrivate`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+
+   You have now created an Interbit account. You can make changes to your
+personal account information and see how the new blocks are added to your
+private account chain with each change.
```
Is there one chain per Interbit account? If so, we should say `...to the private blockchain containing your account information`. If not, we should replace `account` with `accountsPrivate`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+
+## Allow the template app to connect to your private account chain
+
+Now that we have created a user account on the accounts app, we can authorize
```
Capitalize `Accounts`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+the `interbit-template` app to read your name and email from your private
+accounts chain.
+
+1. Open the template app in your browser at `http://localhost:3000`.
```
Capitalize `Template`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+
+1. Open the template app in your browser at `http://localhost:3000`.
+
+1. Open the browser console and verify that the `templatePrivate` chain is
```
Add `developer` between `browser` and `console`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+1. Open the template app in your browser at `http://localhost:3000`.
+
+1. Open the browser console and verify that the `templatePrivate` chain is
+blocking by looking for the `interbit-middleware/CHAIN_BLOCKING` message.
```
Replace `blocking` with `ready for new blocks`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+the **Block Explorer** page and check that the `sharedProfile` property in the
+`templatePrivate` chain state is empty.
+
+1. Make a chain authorization request from the template app to the account
```
Capitalize `Template`. Replace `account` with `Accounts`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+
+1. Make a chain authorization request from the template app to the account
+app. Click the **Connect to Profile** link in the header to navigate to the
+`/cauth/request` page and click the **Request cAuth** button. The browser
```
Add a comma after `page`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+1. Make a chain authorization request from the template app to the account
+app. Click the **Connect to Profile** link in the header to navigate to the
+`/cauth/request` page and click the **Request cAuth** button. The browser
+redirects you to the accounts app and you should see a the following message:
```
Capitalize `Accounts`. Remove the `a` before `the following message:`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+app. Click the **Connect to Profile** link in the header to navigate to the
+`/cauth/request` page and click the **Request cAuth** button. The browser
+redirects you to the accounts app and you should see a the following message:
+Service templatePrivate wants to access the following identity information.
```
Convert this entire sentence to monospace.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+
+   The `templatePrivate` chain is requesting a read join with the
+`accountsPrivate` chain for the `name` and `email` profile fields. Joins can
+be configured so only specific fields are shared.
```
Insert `that` between `so` and `only`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+`accountsPrivate` chain for the `name` and `email` profile fields. Joins can
+be configured so only specific fields are shared.
+
+1. Add any missing profile information that the template app is requesting. If
```
Capitalize `Template`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+be configured so only specific fields are shared.
+
+1. Add any missing profile information that the template app is requesting. If
+you haven't added an email address to your user profile, you will be prompted
```
Replace `will be` with `are`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+block, and that the email address has been added to the profile.
+
+1. Authorize the request for information by clicking the **Accept** button.
+You will be redirected back to the template app and should see a message
```
Replace `will be` with `are`. Capitalize `Template`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+
+1. Authorize the request for information by clicking the **Accept** button.
+You will be redirected back to the template app and should see a message
+stating "Complete the cAuth loop" with a profile identifier of the form
```
Replace:

```
..."Complete the cAuth loop"...
```

with:

```
...`Complete the cAuth flow`...
```

karikushi replied:
In this case, I'm leaving the word `loop` as this is what the Template app actually displays.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+stating "Complete the cAuth loop" with a profile identifier of the form
+`PROFILE-****`.
+
+1. Complete the chain authorization loop by clicking the **Complete cAuth**
```
Replace `loop` with `flow`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+1. Complete the chain authorization loop by clicking the **Complete cAuth**
+button.
+
+   Clicking the button dispatches an action to the `templatePrivate` chain and
```
Replace `Clicking the button` with `The button click`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+   Clicking the button dispatches an action to the `templatePrivate` chain and
+updates the state with the shared profile information.
+
+1. Verify that the `templatePrivate` chain state has the newly shared profile
```
Replace the space between `newly` and `shared` with a hyphen.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+1. Verify that the `templatePrivate` chain state has the newly shared profile
+information. Navigate to the **Block Explorer** and expand the `sharedProfile`
+property. The `name` and `email` fields are populated with the values entered
+from the accounts app.
```
Capitalize `Accounts`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+from the accounts app.
+
+   The `accountsPrivate` and `templatePrivate` chains are now joined. Any
+updates to the user's name or email from the accounts app will be reflected in
```
Capitalize `Accounts`. Replace `will be` with `are automatically`.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,107 @@
+the `templatePrivate` chain.
+
+1. Verify the chain join on the `accountsPrivate` chain. From the
+**Block Explorer** in the accounts app, expand the `shared` property and see
```
Capitalize `Accounts`. Replace `see` with `check`.

btlewan reviewed (changes requested):
A few more minor corrections. Some were missed in the previous update, likely due to GitHub marking comments as outdated because the PR was updated with the review was underway.

Commit:
Make revisions based on PR feedback

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,205 @@
+
+### Generate a set of private and public keys for the `platform-deploy` and `web-auth-endpoint` hypervisors
+
+1. Open a terminal and install `interbit-cli`:
```
This step installs the `interbit-cli` NPM package. The next step talks about the root of the `interbit` repository, which users don't get by installing `interbit-cli`. Does this step need to be replaced with a clone-the-Interbit-repo step, or do we need to keep this step and add a step to clone the repo?

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,205 @@
+for `platform-deploy` and `web-auth-endpoint`, respectively, with:
+
+    ```sh
+    interbit keys --filename platform-deploy-keys.json
````
Should these two keys files, that we create in this step, be considered secrets? If so, let's create the `secrets` folder first, cd into it, and then create these keys files.

karikushi replied:
Definitely, yes. Thanks for catching this 👍

btlewan reviewed (changes requested):
A few comments regarding the run steps.

Commit:
Update instructions for generating interbit keys and running example locally

Commit:
Merge branch 'master' into web-auth-endpoint

Commit:
Update docs based on PR feedback

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,273 @@
+1. In the **Application name** field, give the app a name of your choice,
+e.g. `Interbit Test App`.
+
+1. Fill in the **Homepage URL** field with any URL. Any URL will do since we
```
Replace `will do` with `works`.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,273 @@
+    npm i -g interbit-cli
+    ```
+
+    This will install the [Interbit CLI](/reference/interbit-cli/README.md)
````
Replace `will install` with `installs`.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,273 @@
+    ```
+
+1. Copy the values for the `GITHUB_CLIENT_ID` and `GITHUB_CLIENT_SECRET` fields
+from the GitHub OAuth app you just created (in the open browser tab). These
````
After `...in the open browser tab)` add `, and paste them in the respective fields in the `web-auth-endpoint.sh` file`.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,273 @@
+    ```
+
+1. Copy the public key value from the `platform-deploy-keys.json` file and
+paste it into the `PUB_KEY` variable in `interbit.config.js`.
````
Replace `variable` with `definition`.

btlewan replied:
This step (and the next) seem odd... why go to the trouble of creating keys in a secrets folder, when the keys are being copied into code that isn't ignored by Git? Why don't we suggest editing the line to use `process.env`:

```
const PUB_KEY = process.env.PUBLIC_KEY
```

karikushi replied:
There is an outstanding issue logged for this task at https://github.com/interbit/interbit/issues/250.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,273 @@
+paste it into the `PUB_KEY` variable in `interbit.config.js`.
+
+1. Copy the public key value from the `web-auth-endpoint-keys.json` file and
+paste it into the `WEB_AUTH_PUB_KEY` variable in `interbit.config.js`.
```
Replace `variable` with `definition`.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,273 @@
+    ```
+
+1. Copy the public key value from the `platform-deploy-keys.json` file and
+paste it into the `PUBLIC_KEY` variable in `interbit.config.js`.
````
Replace `variable` with `definition`.

btlewan commented on examples/simple-platform/run.md:
```diff
@@ -0,0 +1,273 @@
+
+   This file must be deleted when new Interbit key pairs are generated (as we
+did in our setup). If there is a pre-existing manifest, the genesis blocks
+are not be overwritten and our new keys do not work. Note that when we
```
Remove `be` from `...not be overwritten...`.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,273 @@
+    cd packages/app-account
+    npm start
+    ```
+
````
We should say that a browser tab opens on port 3025 here. For anyone following along, they might be a bit surprised when the instructions are replaced by the Accounts app.

btlewan replied:
BTW: The Accounts app reported this:

```
Compiled with warnings.

./src/components/ModalAppAccess.js
  Line 47:  Links must not point to "#". Use a more descriptive href or use a button instead  jsx-a11y/href-no-hash

Search for the keywords to learn more about each warning.
To ignore, add // eslint-disable-next-line to the line before.
```

We should probably fix that problem.

karikushi replied:
The ModalAppAccess component isn't used anywhere in the app and is now flagged for removal in Issue #629 of the Interbit repo (https://github.com/interbit/interbit/issues/630).

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,273 @@
+    npm start
+    ```
+
+    Leave this terminal open.
````
We should say that a browser tab opens on port 3000 here. For anyone following along, there's a chance they might be surprised when the instructions are (again) replaced, this time by the Template app.

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,42 @@
+
+Our example has two React applications: `app-account` and `interbit-template`.
+Each app starts its own Interbit node within the browser, making these truly
+"serverless" apps. The Interbit nodes within the browser communicate with the
```
Use italics, instead of quotation marks, when introducing a new term:

```
_serverless_ apps. ...
```

btlewan commented on examples/simple-platform/structure.md:
```diff
@@ -0,0 +1,42 @@
+nodes, static chains, and chain join configurations. See the [Template App
+Walk-through](../template.adoc) for a detailed explanation of the template
+app's chain architecture.
+- Generates a manifest file at
```
Use blank lines to separate multi-line list items from each other.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,108 @@
+
+## Create an account and complete the OAuth flow with GitHub
+
+1. Open the accounts app in your browser at `http://localhost:3025`.
```
This directly follows the `run` steps, so this should be unnecessary. Perhaps the first step here should tell readers to follow the `run` steps first.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,108 @@
+
+1. Open the accounts app in your browser at `http://localhost:3025`.
+
+1. Open the browser's developer tools to the console tab and verify that you
```
What can the user do if the `CHAIN_BLOCKING` status is not included:

```
Download the React DevTools for a better development experience: https://fb.me/react-devtools
interbit-ui-tools.js:11791 interbit-middleware: *rootSaga(): loading interbit
interbit-ui-tools.js:11828 interbit-middleware: *loadInterbitSaga()
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:8464 interbit-middleware: Initializing interbit API
interbit-ui-tools.js:8482 interbit-middleware: Generating key pair
webpackHotDevClient.js:138 ./src/components/ModalAppAccess.js
  Line 47:  Links must not point to "#". Use a more descriptive href or use a button instead  jsx-a11y/href-no-hash
printWarnings @ webpackHotDevClient.js:138
interbit-ui-tools.js:8493 interbit-middleware: Starting interbit hypervisor
interbit-ui-tools.js:8501 interbit-middleware: Creating interbit client
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:12326 interbit-middleware: Connecting to interbit nodes Array(1)0: "localhost:5025"asMutable: ƒ asMutableArray(opts)asObject: ƒ asObject(iterator)concat: ƒ ()filter: ƒ ()flatMap: ƒ flatMap(iterator)getIn: ƒ getIn(path, defaultValue)keys: ƒ ()length: 1map: ƒ ()pop: ƒ ()push: ƒ ()reduce: ƒ ()reduceRight: ƒ ()reverse: ƒ ()set: ƒ arraySet(idx, value, config)setIn: ƒ arraySetIn(pth, value, config)setPrototypeOf: ƒ ()shift: ƒ ()slice: ƒ ()sort: ƒ ()splice: ƒ ()unshift: ƒ ()update: ƒ update(property, updater)updateIn: ƒ updateIn(path, updater)__immutable_invariants_hold: true__proto__: Array(0)
interbit-ui-tools.js:12416 interbit-middleware: Connecting to: localhost:5025
bundle.browser.js:1391 *** NETWORK *** outbound 5TyKzsVKThRUA4SXAAAB is now connected
(anonymous) @ bundle.browser.js:1391
interbit-ui-tools.js:12421 interbit-middleware: Connected to: localhost:5025
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11797 interbit-middleware: *rootSaga(): listening for:  Array(3)
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11871 interbit-middleware: *staticChainsSaga()
interbit-ui-tools.js:12575 interbit-middleware: *loadStaticChains() Object
interbit-ui-tools.js:12907 interbit-middleware: *loadChain() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:8343 interbit-middleware: Connecting chain 'accountsPublic' to store
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11910 interbit-middleware: *privateChainSaga() Object
interbit-ui-tools.js:12609 interbit-middleware: *loadPrivateChain() accountsPrivate
interbit-ui-tools.js:12822 interbit-middleware: *sponsorChain()
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:12907 interbit-middleware: *loadChain() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:8343 interbit-middleware: Connecting chain 'accountsPrivate' to store
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
```

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,108 @@
+
+   You should see a number of chain actions being dispatched in the browser
+console on the Accounts app tab. Specifically, after the action type
+`app-account-my-account/COMPLETE_AUTHENTICATION` is dispatched, the
```
This message doesn't appear for me, even though the account creation appears to be successful.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,108 @@
+1. Verify that your user account information is on the private chain. Navigate
+to the **Block Explorer** page and expand the `profile` object property in the
+`accountsPrivate` chain. If you have not made any changes to your user profile
+information, you should see the that the `name` and `gitHub-identity`
```
`name` contains `anonymous user`, which doesn't seem particularly informative. The `gitHub-identity` key does contain the correct `name` field; why don't we populate `name` from `gitHub-identity/name`? Is that a bug?

karikushi replied:
`anonymous user` is the initial value for the `name` field. The reason we didn't populate it with the GitHub name is that we may support different authentication services later. There are upcoming changes to the account creation and profile editing processes with the Accounts app redesign work.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,108 @@
+the `interbit-template` app to read your name and email from your private
+accounts chain.
+
+1. Open the Template app in your browser at `http://localhost:3000`.
```
If readers are following along, a browser tab is already open.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,108 @@
+
+1. Open the Template app in your browser at `http://localhost:3000`.
+
+1. Open the browser developer console and verify that the `templatePrivate`
```
For me, the `blocking` message does not appear here either:

```
Download the React DevTools for a better development experience: https://fb.me/react-devtools
interbit-ui-tools.js:11791 interbit-middleware: *rootSaga(): loading interbit
interbit-ui-tools.js:11828 interbit-middleware: *loadInterbitSaga()
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:8464 interbit-middleware: Initializing interbit API
interbit-ui-tools.js:8482 interbit-middleware: Generating key pair
interbit-ui-tools.js:8493 interbit-middleware: Starting interbit hypervisor
interbit-ui-tools.js:8501 interbit-middleware: Creating interbit client
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:12326 interbit-middleware: Connecting to interbit nodes Array(1)
interbit-ui-tools.js:12416 interbit-middleware: Connecting to: localhost:5025
bundle.browser.js:1391 *** NETWORK *** outbound QS6Lkbvhy_W2YUAxAAAC is now connected
(anonymous) @ bundle.browser.js:1391
interbit-ui-tools.js:12421 interbit-middleware: Connected to: localhost:5025
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11797 interbit-middleware: *rootSaga(): listening for:  Array(3)
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11871 interbit-middleware: *staticChainsSaga()
interbit-ui-tools.js:12575 interbit-middleware: *loadStaticChains() Object
interbit-ui-tools.js:12907 interbit-middleware: *loadChain() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:8343 interbit-middleware: Connecting chain 'templatePublic' to store
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11910 interbit-middleware: *privateChainSaga() Object
interbit-ui-tools.js:12609 interbit-middleware: *loadPrivateChain() templatePrivate
interbit-ui-tools.js:12822 interbit-middleware: *sponsorChain()
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:12907 interbit-middleware: *loadChain() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:8343 interbit-middleware: Connecting chain 'templatePrivate' to store
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
interbit-ui-tools.js:11576 interbit-middleware: reducer() Object
```

btlewan replied:
It looks like my first run did something, I don't know what, to collapse the reducer status messages down to `Object`. When I run the Accounts app now, I see what the example is describing:

```
...
interbit-middleware: Connecting chain 'accountsPrivate' to store
interbit-ui-tools.js:11576 interbit-middleware: reducer() {type: "interbit-middleware/CHAIN_SUBSCRIBE", payload: {…}}
interbit-ui-tools.js:11576 interbit-middleware: reducer() {type: "interbit-middleware/CHAIN_LOADED", payload: {…}, @@redux-saga/SAGA_ACTION: true}
interbit-ui-tools.js:11576 interbit-middleware: reducer() {type: "interbit-middleware/CHAIN_BLOCKING", payload: {…}, @@redux-saga/SAGA_ACTION: true}
```
I don't think I changed anything, between runs, but all of the apps were stopped and then restarted.

Perhaps we should add a note:

```
If you do not see the `interbit-middleware/CHAIN_BLOCKING` message after 5-10 seconds, reload the page in your browser.
```

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,108 @@
+the **Block Explorer** page and check that the `sharedProfile` property in the
+`templatePrivate` chain state is empty.
+
+1. Make a chain authorization request from the Template app to the Accounts
```
This step comprises multiple steps: separate each user action into its own step.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,108 @@
+`accountsPrivate` chain for the `name` and `email` profile fields. Joins can
+be configured so that only specific fields are shared.
+
+1. Add any missing profile information that the Template app is requesting. If
```
This is *so* weird. We're asking for permission to connect, and we get to modify data before the permission is in place? It raises so many information security questions that I feel we need coverage of why, and the limits imposed.

karikushi replied:
This process is analogous to when an app requests your Facebook data. You get redirected to Facebook, authorize, and go back to the requesting app. All the data modification happens on the Accounts app, not on the Template app.

btlewan commented on examples/simple-platform/user-walk-through.md:
```diff
@@ -0,0 +1,108 @@
+1. Complete the chain authorization flow by clicking the **Complete cAuth**
+button.
+
+   The button click dispatches an action to the `templatePrivate` chain and
```
Nothing else seems to happen. I don't know if it's a bug or not, but the `Complete cAuth` button remains even though the connection appears to be complete. I waited for several minutes for something to appear to happen, bit nothing ever did. It made me wonder if the process failed, or was just incredibly slow. I've tried clicking it several times, and nothing special appears in the console; no errors/warnings at least. Can we change the app to show a success message?

karikushi replied:
There's no visual feedback after clicking the `Complete cAuth` button (terrible UI, I know!!). The feedback you can see is a console message for the action dispatch, and an update to the template and accounts chains with the shared data.

btlewan reviewed (changes requested):
A few minor changes, and a number of questions.

btlewan commented on examples/simple-platform/run.md:
````diff
@@ -0,0 +1,273 @@
+    ```
+
+1. Copy the values for the `GITHUB_CLIENT_ID`, `GITHUB_CLIENT_SECRET`, and
+`GITHUB_REDIRECT_URL` fields from the GitHub OAuth app you just created (in the
````
This step needs updating:
- There are no `GITHUB_*` fields on the GitHub OAuth app page, so we need to tell people to `copy from field X and paste in field Y`.
- There is no `redirect URL` field on the GitHub OAuth page. I used the `Authorization Callback URL`, but I'm not sure if that's what's intended.

karikushi replied:
That's what's intended. I updated the docs so that the copy/paste correspondence is explicitly stated.

btlewan reviewed (changes requested):
One more change required.

Commit:
Merge branch 'master' into web-auth-endpoint

Commit:
Respond to PR feedback

btlewan committed:
Revised platform-deploy ecample:

- Added screenshots to help clarify each step.
- Converted markup to Asciidoc for image sizing.
- Added illustration to help describe chain topology.
- Split config+run file into config and run files.
- Minor wording fixes.
- Applied medium-zoom to screenshots to make them easier to read.

karikushi commented on SUMMARY.md:
```diff
@@ -26,10 +26,11 @@
+    * [Platform Structure](/examples/simple-platform/structure.adoc)
+    * [Configure the Example](/examples/simple-platform/configure.adoc)
+    * [Run the Example Locally](/examples/simple-platform/run.adoc)
+    * [Platform operation](/examples/simple-platform/operation.adoc)
```
Capitalization of `Platform operation` is not consistent with the other titles here

karikushi commented on examples/simple-platform/README.adoc:
```diff
@@ -0,0 +1,57 @@
+-- describing the chain hosting for our apps and the GitHub OAuth flow
+   with the `packages/platform-deploy` and `packages/web-auth-endpoint`
+   hypervisors.
+
```
I compiled the docs and opened them up in the browser. I'm seeing that the sublist items are not formatted correctly. They are on the same line and separated by `--` and without indentation.

karikushi replied:
Looks good now 👍

karikushi commented on examples/simple-platform/README.adoc:
```diff
@@ -0,0 +1,57 @@
+- link:structure.adoc[Platform Structure]
+- link:configure.adoc[Configure the Example]
+- link:run.adoc[Run the Example Locally]
+- link:operation.adoc[Platform operation]
```
Consistent capitalization of `Platform operation`

karikushi commented on examples/simple-platform/structure.adoc:
```diff
@@ -0,0 +1,50 @@
+`platform-deploy` and `web-auth-endpoint` hypervisors, to coordinate
+consensus as actions are dispatched to the respective blockchains.
+
+image:img/platform-deploy.png["The Platform structure", 932, 438,
```
👍 I like the zoomable images

karikushi commented on examples/simple-platform/structure.adoc:
```diff
@@ -0,0 +1,50 @@
+  `packages/interbit-template/interbit.config.js`. These files specify
+  peer nodes, static chains, and chain join configurations. See the
+  link:../template.adoc[Template App Walk-through] for a detailed
+  explanation of the template app's chain architecture.
```
Oops, missed a capitalization of `template app` here.

karikushi commented on examples/simple-platform/configure.adoc:
```diff
@@ -0,0 +1,251 @@
+
+. Visit link:https://github.com/[github.com] and log into your account.
++
+image:img/screen-github-home.png["The GitHub home page", 1280, 311,
```
👍 Yay, screenshots

karikushi commented on examples/simple-platform/configure.adoc:
```diff
@@ -0,0 +1,251 @@
+[source,bash]
+cd secrets
+
+. Generate Interbit key pairs
```
We're missing a period at the end of `Generate Interbit key pairs`

karikushi commented on examples/simple-platform/configure.adoc:
```diff
@@ -0,0 +1,251 @@
+from this tab in subsequent steps.
+
+
+== Clone the `interbit` repository and install `interbit-cli`
```
This is a minor point, but the font size for `interbit` and `interbit-cli` in these subheadings looks larger than the regular subheading text. It makes the headings feel off balance, visually.

karikushi commented on examples/simple-platform/configure.adoc:
```diff
@@ -0,0 +1,251 @@
+. Copy the `Client Secret` value from the GitHub browser tab and paste
+  it into the `GITHUB_CLIENT_SECRET` definition in `platform-deploy.sh`.
+
+. In `Authorization callback URL` definition in `platform-deploy.sh`,
```
It's the `GITHUB_REDIRECT_URL` definition that gets the specified value. This value should be the same as the `Authorization callback URL` specified when you created the GitHub OAuth app.

karikushi commented on examples/simple-platform/configure.adoc:
```diff
@@ -0,0 +1,251 @@
+== Configuration complete
+
+At this point, all of the apps are configured. See the
+link:run.adoc[next section] to see how to start all four applications.
```
Can we specifically say Template app, Accounts app, `platform-deploy` node, and `web-auth-endpoint` node since the last two are not apps?

karikushi commented on examples/simple-platform/run.adoc:
```diff
@@ -0,0 +1,109 @@
+= Run the Example Locally
+
+This section describes how to start all four apps that are used in this
```
Can we specifically say Template app, Accounts app, `platform-deploy` node, and `web-auth-endpoint` node since the last two are not apps?

karikushi commented on examples/simple-platform/run.adoc:
```diff
@@ -0,0 +1,109 @@
++
+IMPORTANT: Leave this terminal open.
+
+. Wait 3 minutes.
```
It might be a good idea to include a sample of what's printed to the terminal when the chains have started up correctly. I *think* this is when you see `Applying covenant XXXX to templateControl (YYYY) JOINING....` for example.

karikushi commented on examples/simple-platform/run.adoc:
```diff
@@ -0,0 +1,109 @@
+
+All four apps are now running. See the link:operation.adoc[next section]
+for the steps to create an account by authenticating with GitHub, and to
+authorize another app to connect to your account chain.
```
Two apps and two nodes

karikushi commented on examples/simple-platform/operation.adoc:
```diff
@@ -0,0 +1,226 @@
+role="frame"]
+
+
+== Allow the template app to connect to your private account chain
```
Capitalize `template app`

karikushi commented on examples/simple-platform/operation.adoc:
```diff
@@ -0,0 +1,226 @@
++
+image:img/screen-accounts_app-be-shared-connected.png["The Block
+Explorer screen in the Accounts app, showing the shared profile now
+contains populated data", 1280, 581, role="frame"]
```
Could we add a conclusion for this section and the entire platform example? The other sections wrap up nicely with short statements about what was completed in the section and what's to come in the next section. This feels like an abrupt ending to the example. (Sorry, I should have written one in my first attempts at these docs).

btlewan committed:
Apply PR feedback.

karikushi commented on examples/simple-platform/operation.adoc:
```diff
@@ -224,3 +224,17 @@ chain.
+- how those coordinated nodes and apps work in practice.
+
+Now that you have a working example, you have a foundation to tinker and
+learn further.
```
😸

karikushi commented on examples/simple-platform/run.adoc:
```diff
@@ -34,13 +34,21 @@ Account and Template apps' browser nodes.
+Once you see the following line, `platform-deploy` is fully initialized
+and you can move on to the next step:
++
+`ROOT SAGA: watching for app-account-github-kyc/OAUTH_CALLBACK_SAGA`
```
👍

karikushi reviewed (commented):
LGTM

Commit:
Merge branch 'master' into web-auth-endpoint

btlewan committed:
Fix outdated link.
